include/rtc/messagehandlerelement.hpp

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright (c) 2020 Filip Klembara (in2core)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef RTC_MESSAGE_HANDLER_ELEMENT_H
+#define RTC_MESSAGE_HANDLER_ELEMENT_H
+
+#if RTC_ENABLE_MEDIA
+
+#include "include.hpp"
+#include "message.hpp"
+#include "rtp.hpp"
+
+namespace rtc {
+
+using ChainedMessagesProduct = std::shared_ptr<std::vector<binary_ptr>>;
+
+RTC_CPP_EXPORT ChainedMessagesProduct make_chained_messages_product();
+RTC_CPP_EXPORT ChainedMessagesProduct make_chained_messages_product(message_ptr msg);
+
+/// Ougoing messages
+struct RTC_CPP_EXPORT ChainedOutgoingProduct {
+	ChainedOutgoingProduct(ChainedMessagesProduct messages, std::optional<message_ptr> control = nullopt);
+	const ChainedMessagesProduct messages;
+	const std::optional<message_ptr> control;
+};
+
+/// Ougoing response to incoming messages
+struct RTC_CPP_EXPORT ChainedOutgoingResponseProduct {
+	ChainedOutgoingResponseProduct(std::optional<ChainedMessagesProduct> messages = nullopt, std::optional<message_ptr> control = nullopt);
+	const std::optional<ChainedMessagesProduct> messages;
+	const std::optional<message_ptr> control;
+};
+
+/// Incoming messages with response
+struct RTC_CPP_EXPORT ChainedIncomingProduct {
+	ChainedIncomingProduct(std::optional<ChainedMessagesProduct> incoming = nullopt, std::optional<ChainedMessagesProduct> outgoing = nullopt);
+	const std::optional<ChainedMessagesProduct> incoming;
+	const std::optional<ChainedOutgoingResponseProduct> outgoing;
+};
+
+/// Incoming control messages with response
+struct RTC_CPP_EXPORT ChainedIncomingControlProduct {
+	ChainedIncomingControlProduct(message_ptr incoming, std::optional<ChainedOutgoingResponseProduct> outgoing = nullopt);
+	const std::optional<message_ptr> incoming;
+	const std::optional<ChainedOutgoingResponseProduct> outgoing;
+};
+
+/// Chainable handler
+class RTC_CPP_EXPORT MessageHandlerElement: public std::enable_shared_from_this<MessageHandlerElement> {
+	std::optional<std::shared_ptr<MessageHandlerElement>> upstream = nullopt;
+	std::optional<std::shared_ptr<MessageHandlerElement>> downstream = nullopt;
```

Comment:
This is just my personal preference (I can change it if you want). By wrapping pointer in optional I declare that value of pointer won't be `nullptr` so there should not be need for `nullptr` check (I don't like idea where I work with shared pointers and need to think if it can be `nullptr` or not).

Reply:
I see! Since in other places `message_ptr` may be null, I think it would be great if you can change to do the same for the sake of consistency.

Reply:
done